src/common/Locator.c

```diff
@@ -0,0 +1,230 @@
+	for (size_t i = 4; i < 6; ++i) {
+		locator->grid[i] = to_ascii_lowercase(locator->grid[i]);
+	}
+
+	// translate space to NUL, which will truncate the grid square
+	for (size_t i = 0; i < sizeof(locator->grid); ++i) {
+		if (locator->grid[i] == ' ') {
+			locator->grid[i] = 0;
+		}
+	}
+
+	return locator_validate_grid(locator);
+}
+
+ARDOP_MUSTUSE locator_err
+locator_validate_grid(const Locator* locator) {
+	size_t len = strnlen(locator->grid, sizeof(locator->grid));
+	if (len % 2 != 0) {
+		return LOCATOR_ERR_FMT_LENGTH;
+	}
+
+	for (size_t i = 0; i < len; ++i) {
+		switch (i/2) {
+			case 0:
+				// field A – R
+				if (! is_ascii_alpha_range(locator->grid[i], 'R'))
+					return LOCATOR_ERR_FMT_FIELD;
+				break;
+			case 1:
+				// square 0 – 9
+				if (! is_ascii_digit(locator->grid[i]))
+					return LOCATOR_ERR_FMT_SQUARE;
+				break;
+			case 2:
+				// subsquare A – X
+				if (! is_ascii_alpha_range(locator->grid[i], 'X'))
+					return LOCATOR_ERR_FMT_SUBSQUARE;
+				break;
+			default:
+				// extended square, 0 – 9
+				if (! is_ascii_digit(locator->grid[i]))
+					return LOCATOR_ERR_FMT_EXTSQUARE;
+				break;
+		}
+	}
+
+	return LOCATOR_OK;
+}
+
+void locator_init(Locator* locator) {
+	memset(locator, 0, sizeof(*locator));
+}
+
+locator_err locator_from_str(const char* str, Locator* locator) {
+	const char* inp = str ? str : "";
+	size_t len = strlen(inp);
+	return locator_from_str_slice(inp, len, locator);
+}
+
+locator_err locator_from_str_slice(const char* str, size_t len, Locator* locator) {
+	locator_err out = LOCATOR_OK;
+
+	// empty grid is OK
+	if (len == 0)
+		return out;
+
+	// copy string, checking for truncation
+	int ck = snprintf(locator->grid, sizeof(locator->grid), "%.*s", (int)len, str);
+	if (ck <= 0 || ck >= sizeof(locator->grid)) {
+		locator_init(locator);
+		return LOCATOR_ERR_TOOLONG;
+	}
+
+	// validate and compress
+	out = locator_compress(locator, &locator->wire);
+	if (out) {
+		locator_init(locator);
+		return out;
+	}
+
+	// canonicalize by uncompressing
+	out = locator_uncompress(&locator->wire, locator);
+	if (out) {
+		locator_init(locator);
+		return out;
+	}
+
+	return out;
+}
+
+const Packed6* locator_as_bytes(const Locator* locator) {
+	if (!! locator && locator_is_populated(locator)) {
+		return &(locator->wire);
+	} else {
+		return &LOCATOR_NOGS[1];
+	}
+}
+
+locator_err locator_from_bytes(const uint8_t bytes[PACKED6_SIZE], Locator* locator) {
+	locator_init(locator);
+
+	// read Packed6
+	packed6_from_bytes(bytes, &locator->wire);
+
+	// decompress and validate
+	locator_err out = locator_uncompress(&locator->wire, locator);
+	if (out) {
+		locator_init(locator);
+		return out;
+	}
+
+	return out;
+}
+
+bool locator_is_populated(const Locator* locator) {
+	return locator->grid[0] != '\0';
+}
+
+const char* locator_strerror(locator_err err) {
+	const static char* MSGS[] = {
+		"unknown error",
+		"length exceeded",
+		"locator must be 2, 4, 6, or 8 characters",
```

Comment:
This is a change in behavior.  Previously, a 2-character gridsquare was not accepted.  However, since not providing a gridsquare is permissible, I cannot see a reason to reject a very coarse gridsquare.